R/stacking.R

```diff
@@ -0,0 +1,1626 @@
+#' @param include.stacked.data.set.in.output Whether to include the stacked
+#'   data set in the output.
+#' @param include.original.case.variable Whether to include the \code{original_case}
+#'   variable in the stacked data set.
+#' @param include.observation.variable Whether to include the \code{observation}
+#'   variable in the stacked data set.
+#' @return A list with the following elements:
+#' \itemize{
+#'   \item \code{stacked.data.set.metadata} A list containing metadata on the
+#'     the stacked data set such as variable names, labels etc.
+#'   \item \code{unstackable names} A list of character vectors containing
+#'     names of the variables that could not be stacked using common labels due
+#'     to mismatching types or value attributes.
+#'   \item \code{omitted.variables} A character vector of omitted variables.
+#'   \item \code{omitted.stacked.variables} A character vector of omitted
+#'     stacked variables.
+#'   \item \code{common.labels} A character vector of common labels used for
+#'     stacking.
+#'   \item \code{is.saved.to.cloud} Whether the stacked data set was saved to
+#'     the Displayr cloud drive.
+#'  }
+#' @export
+StackData <- function(input.data.set.name,
+                      stacked.data.set.name = NULL,
+                      stack.with.common.labels = "Automatically",
+                      reference.variables.to.stack = NULL,
+                      manual.common.labels = NULL,
+                      specify.by = "Variable",
+                      manual.stacking = NULL,
+                      variables.to.omit = NULL,
+                      write.data.set = TRUE,
+                      include.stacked.data.set.in.output = FALSE,
+                      include.original.case.variable = TRUE,
+                      include.observation.variable = TRUE)
+{
+    input.data.set <- readDataSets(input.data.set.name, 1)[[1]]
+    variables.to.omit <- splitVariablesToOmitText(variables.to.omit)
+    omitted.variables <- parseVariablesToOmit(variables.to.omit,
+                                              names(input.data.set),
+                                              FALSE)
+
+    input.data.set <- omitVariablesFromDataSet(input.data.set,
+                                               omitted.variables)
+    input.data.set.metadata <- metadataFromDataSet(input.data.set,
+                                                   input.data.set.name)
+
+    common.labels.list <- commonLabels(manual.common.labels,
+                                       stack.with.common.labels,
+                                       input.data.set.metadata,
+                                       reference.variables.to.stack)
+
+    stacking.groups <- stackWithCommonLabels(common.labels.list,
+                                             input.data.set.metadata)
+    stacking.groups <- stackManually(stacking.groups, manual.stacking,
+                                     specify.by, input.data.set.metadata)
```

Comment:
The way you pass down stacking.groups into stackManually(), and then merge the two internally, is confusing.  It would be better to create stacking.groups.from.labels and stacking.groups.manual and then merge the two at the top level.  That simplifies stackManually() and makes it easier to unit test.  It also surfaces a critical part of the algorithm (when they are merged).

Reply:
I have refactored the code as you described and created `mergeCommonLabelAndManualStackingGroups`